File: LowestCommonAncestor_BST_Recursive.java
// Time Complexity : O(log(n)) --> where n is number of elements in the tree
// Space Complexity : O(h) --> where h is the height of tree
// Did this code successfully run on Leetcode (235):
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

// Recursive Solution
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        // logic
        if (p.val < root.val && q.val < root.val) return lowestCommonAncestor(root.left, p, q);
        else if (p.val > root.val && q.val > root.val) return lowestCommonAncestor(root.right, p, q);
        else return root;
    }
}

File: lowest_common_ancestor_BST_recursive.py
# Time Complexity : O(H) or O(log N) whenre H is height of the BST and N is number of Nodes in the tree
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if p.val < root.val and q.val < root.val:
            return self.lowestCommonAncestor(root.left, p, q)
        elif p.val > root.val and q.val > root.val:
            return self.lowestCommonAncestor(root.right, p, q)
        
        return root